### PR TITLE
Testing Components with TestBed

### PR DESCRIPTION
**Pull request by justindujardin**
This adds support for jasmine/mocha angular testing using the TestBed class exported by Angular.

Fixes: #752, #479, #606

- [x] Zone.js loads appropriate patches for jasmine/mocha
- [x] Existing tests switched to TestBed from TestApp
- [x] TestBed specs passing
  - [x] ListView
  - [x] ModalDialog
  - [x] Platform Filter Components
  - [x] Third Party Components
  - [x] Renderer E2E
  - [x] Renderer createElement
  - [x] Renderer attach/detach
  - [x] Renderer lifecycle
  - [x] DetachedLoader

### CODE REVIEW

**Comment by justindujardin** [comment]
@vakrilov Most of the groundwork is done for TestBed. I'm in the process of updating the existing tests that use `TestApp` to use `TestBed` instead. 

There are two tests that currently fail that I could use feedback on. I left TODO comments describing the issues:
- https://github.com/NativeScript/nativescript-angular/pull/1061/files#diff-a4db03e8331a4981dfb156a14c7cbe5cR55
- https://github.com/NativeScript/nativescript-angular/pull/1061/files#diff-21b5ac960e580c34c4a5ee9286f0eddaR42

**Comment by justindujardin** [comment]
I think this is ready for review now. 👍 

Notes
- I intended to write specs to verify TestBed, but it seemed a little weird testing the testing setup, and it gets pretty thorough testing from the existing specs in the `nativescript-angular` codebase.
- I also intended to remove the `TestApp` class, but there are router tests that use it and assert about the app. It seems an appropriate fixture for that test, so it gets to hang around.

**Comment by vakrilov** [comment]
Kudos for this massive PR!!! 
Running some CI builds:

Run ci
e2e

**Comment by justindujardin** [comment]
@vakrilov I noticed some of the Jenkins tests failed when you ran the ci job, but I can't access the logs. I'm happy to debug any issues if you can help me understand how to reproduce them locally. Whenever you have some time, I'm not in a rush.

**Comment by vchimev** [comment]
renderer-android

**Comment by hypery2k** [comment]
is possible to also back port this one to nativescript-angular 4.x?

**Comment by vakrilov** [comment]
Run ci
e2e

File: nativescript-angular/testing/src/util.ts [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+ * Return a promise that resolves after (durationMs) milliseconds
+ */
+export function promiseWait(durationMs: number) {
+    return () => new Promise((resolve) => setTimeout(() => resolve(), durationMs));
+}
+
+/**
+ * Perform basic TestBed environment initialization. Call this once in the main entry point to your tests.
+ */
+export function nTestBedInit() {
```
[the remark this answers is not in the document]
Comment (justindujardin) [reply nativescript-angular/testing/src/util.ts]: Sure, what about:

```
export function testingRootView(): View
export function promiseWait(durationMs: number)
```

**Comment by hypery2k** [comment]
@vakrilov can you retrigger the Jenkins build?

**Comment by sis0k0** [comment]
run ci

**Comment by justindujardin** [comment]
Hi, there are some problems with bootstrapping the test app in @angular ~5.1.0 that I'm debugging. I'll update here when I think it's worth running the CI build again.

**Comment by sis0k0** [comment]
@justindujardin, you can target Angular 5.2.0, it'll make rebasing easier.

**Comment by hypery2k** [comment]
@justindujardin Any chance for you to update your PR?

**Comment by abdulhaq-e** [comment]
Hey @justindujardin , any updates on this PR ?

Thanks a lot, great work.

**Comment by hypery2k** [comment]
@abdulhaq-e I've updated his branch, see #1175

**Comment by abdulhaq-e** [comment]
@hypery2k Nice! have your tried it out?

Also, why haven't the tests completed in your PR?

**Comment by Smiley01987** [comment]
When will this be available?

**Comment by justindujardin** [comment]
Sorry, it looks like I won't have free time to continue work on this problem in the short term.

At this point perhaps @vakrilov can find a team member to resolve the conflicts and debug the `tns test ios` startup failure that happens with angular 5 (a guess, it didn't happen in the original PR for 4.x)?

The code and tests here should still be 100% okay.

**Comment by ndcunningham** [comment]
any update on this?

**Comment by AresDev** [comment]
This is really a need, any update for this?

**Comment by vakrilov** [comment]
test

**Comment by vakrilov** [comment]
Closing in favor of  #1175

**Pull request closed without merging**